sdk/python/feast/pyspark/launchers/aws/emr.py

```diff
@@ -0,0 +1,244 @@
+
+        Args:
+            output_file_uri (str): Uri to the historical feature retrieval job output file.
+        """
+        super().__init__(emr_client, job_ref)
+        self._output_file_uri = output_file_uri
+
+    def get_output_file_uri(self, timeout_sec=None):
+        _wait_for_job_state(
+            self._emr_client, self._job_ref, TERMINAL_STEP_STATES, timeout_sec
+        )
+        return self._output_file_uri
+
+
+class EmrIngestionJob(EmrJobMixin, IngestionJob):
+    """
+    Ingestion job result for a Dataproc cluster
+    """
+
+    def __init__(self, emr_client, job_ref: EmrJobRef):
+        super().__init__(emr_client, job_ref)
+
+
+class EmrClusterLauncher(JobLauncher):
+    """
+    Submits jobs to an existing or new EMR cluster. Requires boto3 as an additional dependency.
+    """
+
+    _existing_cluster_id: Optional[str]
+    _new_cluster_template: Optional[Dict[str, Any]]
+    _staging_location: str
+    _emr_log_location: str
+    _region: str
+
+    def __init__(
+        self,
+        *,
+        region: str,
+        existing_cluster_id: Optional[str],
+        new_cluster_template_path: Optional[str],
+        staging_location: str,
+        emr_log_location: str,
+    ):
+        """
+        Initialize a dataproc job controller client, used internally for job submission and result
+        retrieval.
+
+        Args:
+            cluster_name (str):
+                Dataproc cluster name.
+            staging_location (str):
+                GCS directory for the storage of files generated by the launcher, such as the pyspark scripts.
+            region (str):
+                Dataproc cluster region.
+            project_id (str:
+                GCP project id for the dataproc cluster.
+        """
+
+        assert existing_cluster_id or new_cluster_template_path
+
+        self._existing_cluster_id = existing_cluster_id
+        if new_cluster_template_path:
+            self._new_cluster_template = _load_new_cluster_template(
+                new_cluster_template_path
+            )
+        else:
+            self._new_cluster_template = None
+
+        self._staging_location = staging_location
+        self._emr_log_location = emr_log_location
+        self._region = region
+
+    def _emr_client(self):
+        return boto3.client("emr", region_name=self._region)
+
+    def _submit_emr_job(self, step: Dict[str, Any]) -> EmrJobRef:
+        """
+        Submit EMR job using a new or existing cluster.
+
+        Returns a tuple of JobId, StepId
+        """
+
+        emr = self._emr_client()
+
+        if self._existing_cluster_id:
+            step["ActionOnFailure"] = "CONTINUE"
+            step_ids = emr.add_job_flow_steps(
+                JobFlowId=self._existing_cluster_id, Steps=[step],
+            )
+            return EmrJobRef(self._existing_cluster_id, step_ids["StepIds"][0])
+        else:
```

Comment:
Personal preference, but I prefer to not have an "else" if the "if" returns.

Reply:
i'll leave it as is here, if it was a "short circuit" early return I'd agree, but here it is symmetric since two branches are equivalent alternative code paths (using an existing cluster `else` using a new cluster)